.github/workflows/build_packages.yaml

```
@@ -16,148 +16,115 @@ jobs:
+          # discrepancies in the images' environments affecting even basic things like
+          # repo checkouts.
           run: |
-            docker run -i --name emqtt-$SYSTEM-build -v $(pwd):/emqtt emqx/build-env:$ERL_OTP-$SYSTEM /bin/bash -c "cd /emqtt && .github/workflows/script/build.sh"
-            cd _packages && for var in $(ls); do sudo bash -c "echo $(sha256sum $var | awk '{print $1}') > $var.sha256"; done && cd -
-        - uses: actions/upload-artifact@v1
+            docker run --rm --name emqtt-${{ matrix.os }}-build -v $(pwd):/emqtt --workdir /emqtt ${BUILDER_IMAGE} /bin/bash -c '
+              git config --global --add safe.directory "${PWD}"
+              .github/workflows/script/build.sh
+            '
+            sudo bash -c '
+              cd _packages
+              for var in $(ls); do echo $(sha256sum $var | awk "{print $1}") > $var.sha256; done
+            '
+        - uses: actions/upload-artifact@v3
           with:
-            name: packages-${{ matrix.os }}
-            path: _packages/.
+            name: packages
+            path: _packages/*
 
-    build-on-mac-erlang-24:
+    build-on-mac:
         runs-on: macos-latest
 
-        steps:
-        - uses: actions/checkout@v1
-        - name: prepare
-          run: |
-            /usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
-            brew install curl zip unzip gnu-sed erlang@24 coreutils
-            echo 'export PATH="/usr/local/opt/unzip/bin:$PATH"' >> ~/.bashrc
-        - name: build
-          run: |
-            export PATH="/usr/local/opt/erlang@24/bin:$PATH"
-            echo "$PATH"
-            erl -noshell -eval 'io:format(erlang:system_info(otp_release)), halt(0).'
-            .github/workflows/script/build.sh
-            pkg=emqtt-macos-$(git describe --tags --always).zip
-            openssl dgst -sha256 _packages/$pkg | awk '{print $2}' > _packages/$pkg.sha256
-        - uses: actions/upload-artifact@v1
-          with:
-            name: packages-mac
-            path: _packages/.
-
-    build-on-mac-erlang-23:
-        runs-on: macos-latest
+        strategy:
+          fail-fast: false
+          matrix:
+            otp:
+              - 24
+              - 25
 
         steps:
-        - uses: actions/checkout@v1
+        - uses: actions/checkout@v3
         - name: prepare
           run: |
-            /usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
-            brew install curl zip unzip gnu-sed erlang@23 coreutils
-            echo 'export PATH="/usr/local/opt/unzip/bin:$PATH"' >> ~/.bashrc
+            # See https://github.com/actions/setup-python/issues/577
+            # brew update
+            brew install curl zip unzip gnu-sed erlang@${{ matrix.otp }} coreutils
+            echo "/usr/local/opt/erlang@${{ matrix.otp }}/bin" >> $GITHUB_PATH
```

to ensure compatibility with apple silicon later
```suggestion
            echo "$(brew --prefix)/opt/erlang@${{ matrix.otp }}/bin" >> $GITHUB_PATH
```